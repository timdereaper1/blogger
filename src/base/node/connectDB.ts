import { MongoClient } from 'mongodb';

let cached: { conn: MongoClient | null; promise: Promise<MongoClient> | null };

if (!cached) {
	cached = global.mongo = { conn: null, promise: null };
}

const environment = {
	development: {
		database: 'blogger',
		url: process.env.MONGO_URL,
	},
	test: {
		database: 'test_blogger',
		url: process.env.MONGO_URL,
	},
	production: {
		database: 'blogger',
		url: '',
	},
};

export function getDatabaseConfig() {
	const config = environment[process.env.NODE_ENV];
	return config;
}

export async function createDatabaseConnection() {
	if (cached.conn) return cached.conn;

	if (!cached.promise) {
		const config = getDatabaseConfig();
		cached.promise = MongoClient.connect(config.url);
		cached.conn = await cached.promise;
		return cached.conn;
	}
}
