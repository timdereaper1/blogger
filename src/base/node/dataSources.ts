import { Db } from 'mongodb';
import { UsersRepository, UsersRepositoryInterface } from './repositories/usersRepository';

export interface DataSourcesInterface {
	users: UsersRepositoryInterface;
}

export function DataSources(db: Db): DataSourcesInterface {
	const users = UsersRepository(db);

	return Object.freeze<DataSourcesInterface>({
		users,
	});
}
