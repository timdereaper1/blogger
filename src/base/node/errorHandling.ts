import { ApolloError } from 'apollo-server-micro';
import { processErrorToErrorLogs } from './logging';

export function processRequestError(error: any, defaultMessage?: string | ApolloError) {
	if (error.isCustom && defaultMessage) throwApolloErrorOrErrorMessage(defaultMessage);
	if (error.isCustom) throw error;
	processErrorToErrorLogs(error);
	throw new Error('Internal server error');
}

export function throwApolloErrorOrErrorMessage(error: string | ApolloError) {
	if (typeof error === 'string') throw new Error(error);
	throw error;
}
