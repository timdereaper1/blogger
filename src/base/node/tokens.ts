import { decode, sign, verify } from 'jsonwebtoken';

export function createAuthenticationToken(args: string) {
	return sign(args, 'secret-key-used-for-tokens');
}

export function getAuthenticatedUserIdInToken(token: string): string | null {
	const isTokenValid = verifyAuthenticationToken(token);
	if (!isTokenValid) return null;
	const decoded = decode(token);
	if (!decoded) return null;
	if (typeof decoded === 'object' && 'id' in decoded) return decoded.id;
	return decoded as any;
}

export function createExpiryAuthenticationToken(args: string, expiresIn: string | number = '1h') {
	return sign({ id: args }, 'secret-key-used-for-tokens', {
		expiresIn,
	});
}

export function verifyAuthenticationToken(token: string) {
	try {
		verify(token, 'secret-key-used-for-tokens');
		return true;
	} catch (error) {
		return false;
	}
}
