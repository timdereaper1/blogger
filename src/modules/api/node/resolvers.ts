import {
	signUpAccountMutation,
	verifyAndSendPasswordResetEmailMutation,
	verifyLoginCredentialsMutation,
} from 'src/modules/authentication/node/authenticationResolvers';

export const resolvers = {
	Query: {
		hello() {
			return 'Hello world';
		},
	},
	Mutation: {
		verifyCredentials: verifyLoginCredentialsMutation,
		signUpAccount: signUpAccountMutation,
		verifyAndSendPasswordResetEmail: verifyAndSendPasswordResetEmailMutation,
	},
};
