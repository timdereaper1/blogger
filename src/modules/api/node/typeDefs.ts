import { gql } from 'apollo-server-micro';

export const typeDefs = gql`
	type User {
		id: String!
		name: String!
		email: String!
		token: String
		privileges: [String!]!
	}

	type Success {
		message: String!
		success: Boolean!
	}

	input UserLoginCredentials {
		email: String!
		password: String!
	}

	input UserSignUpCredentials {
		password: String!
		email: String!
		name: String!
		privileges: [String!]!
	}

	input UserPasswordResetCredentials {
		email: String!
	}

	type Query {
		hello: String
	}

	type Mutation {
		verifyCredentials(data: UserLoginCredentials!): User
		signUpAccount(data: UserSignUpCredentials!): User
		verifyAndSendPasswordResetEmail(data: UserPasswordResetCredentials!): Success
	}
`;
