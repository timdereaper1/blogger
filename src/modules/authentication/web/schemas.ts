import { gql } from '@apollo/client';

export const AUTHENTICATED_USER_FRAGMENT = gql`
	fragment AuthenticatedUser on User {
		id
		name
		email
		token
		privileges
	}
`;

export const SIGN_UP_ACCOUNT_MUTATION = gql`
	mutation signUpAccount($data: UserSignUpCredentials!) {
		signUpAccount(data: $data) {
			...AuthenticatedUser
		}
	}

	${AUTHENTICATED_USER_FRAGMENT}
`;

export const SIGN_IN_ACCOUNT_MUTATION = gql`
	mutation verifyCredentials($data: UserLoginCredentials!) {
		verifyCredentials(data: $data) {
			...AuthenticatedUser
		}
	}

	${AUTHENTICATED_USER_FRAGMENT}
`;

export const VERIFY_AND_SEND_PASSWORD_RESET_EMAIL_MUTATION = gql`
	mutation verifyAndSendPasswordResetEmail($data: UserPasswordResetCredentials!) {
		verifyAndSendPasswordResetEmail(data: $data) {
			message
			success
		}
	}
`;

export const RESET_USER_PASSWORD_MUTATION = gql`
	mutation resetUserPassword($data: ResetPasswordCredentials!) {
		resetUserPassword(data: $data) {
			message
			success
		}
	}
`;
