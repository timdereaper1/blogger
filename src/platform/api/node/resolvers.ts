import { loginGraphQLMutationResolvers } from 'src/platform/login/node/loginGraphQLResolver';

export const resolvers = {
	Query: {
		hello() {
			return 'Hello world';
		},
	},
	Mutation: Object.assign({}, loginGraphQLMutationResolvers),
};
