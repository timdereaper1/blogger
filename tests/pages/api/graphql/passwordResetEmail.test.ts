import { ApolloServer } from 'apollo-server-micro';
import { Collection, Db } from 'mongodb';
import { DataSources } from 'src/base/node/dataSources';
import { createAuthenticationToken } from 'src/base/node/tokens';
import { resolvers } from 'src/modules/api/node/resolvers';
import { typeDefs } from 'src/modules/api/node/typeDefs';
import { VERIFY_AND_SEND_PASSWORD_RESET_EMAIL_MUTATION } from 'src/modules/authentication/web/schemas';
import { authToken } from 'tests/fixtures/token';
import { getDatabaseUserFixture } from 'tests/fixtures/users';

jest.mock('src/base/node/services/mailing');
jest.mock('src/base/node/tokens');

const mockedToken = createAuthenticationToken as jest.MockedFunction<
	typeof createAuthenticationToken
>;

describe('graphql: send password reset email', () => {
	const collection: Partial<Record<keyof Collection, jest.Mock>> = {
		findOne: jest.fn(),
	};

	const db: Partial<Record<keyof Db, jest.Mock>> = {
		collection: jest.fn().mockReturnValue(collection),
	};

	const sources = DataSources(db as any);

	const server = new ApolloServer({
		typeDefs,
		resolvers,
		dataSources: (): any => sources,
	});

	it('should send an email to user to reset their password', async () => {
		const user = await getDatabaseUserFixture();
		collection.findOne.mockResolvedValueOnce(user);
		mockedToken.mockReturnValueOnce(authToken);
		const response = await server.executeOperation({
			query: VERIFY_AND_SEND_PASSWORD_RESET_EMAIL_MUTATION,
			variables: { data: { email: 'john.doe@gmail.com' } },
		});
		expect(response).toMatchSnapshot();
	});
});
