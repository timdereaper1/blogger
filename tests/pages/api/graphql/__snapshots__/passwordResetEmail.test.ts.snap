// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`send password reset email should send an email to user to reset their password 1`] = `
Object {
  "data": Object {
    "verifyAndSendPasswordResetEmail": Object {
      "message": "Reset password email has been sent to John Doe",
      "success": true,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
